File: sdks/go/crypto.go
package xagentauth

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/base64"
	"encoding/hex"
)

// HmacSHA256Hex computes the HMAC-SHA256 of the message using the secret
// and returns the result as a hex-encoded string.
func HmacSHA256Hex(message, secret string) string {
	h := hmac.New(sha256.New, []byte(secret))
	h.Write([]byte(message))
	return hex.EncodeToString(h.Sum(nil))
}

// TimingSafeEqual performs a constant-time comparison of two strings
// to prevent timing attacks.
func TimingSafeEqual(a, b string) bool {
	return subtle.ConstantTimeCompare([]byte(a), []byte(b)) == 1
}

// RandomBytes generates cryptographically secure random bytes.
func RandomBytes(length int) []byte {
	buf := make([]byte, length)
	_, err := rand.Read(buf)
	if err != nil {
		panic("crypto/rand failed: " + err.Error())
	}
	return buf
}

// ToHex converts bytes to a lowercase hex string.
func ToHex(data []byte) string {
	return hex.EncodeToString(data)
}

// FromHex converts a hex string to bytes.
func FromHex(hexStr string) ([]byte, error) {
	return hex.DecodeString(hexStr)
}

// SHA256Hex computes the SHA-256 hex digest of the data.
func SHA256Hex(data []byte) string {
	h := sha256.Sum256(data)
	return hex.EncodeToString(h[:])
}

// SHA256Bytes computes the SHA-256 hash and returns raw bytes.
func SHA256Bytes(data []byte) []byte {
	h := sha256.Sum256(data)
	return h[:]
}

// HmacSHA256Bytes computes HMAC-SHA256 of raw bytes, returning raw bytes.
func HmacSHA256Bytes(key, message []byte) []byte {
	h := hmac.New(sha256.New, key)
	h.Write(message)
	return h.Sum(nil)
}

// Base64Encode encodes bytes to standard base64 string.
func Base64Encode(data []byte) string {
	return base64.StdEncoding.EncodeToString(data)
}

// Base64Decode decodes a standard base64 string to bytes.
func Base64Decode(s string) ([]byte, error) {
	return base64.StdEncoding.DecodeString(s)
}

// GenerateID generates a challenge ID like "ch_<32 hex chars>".
func GenerateID() string {
	return "ch_" + hex.EncodeToString(RandomBytes(16))
}

// GenerateSessionToken generates a session token like "st_<48 hex chars>".
func GenerateSessionToken() string {
	return "st_" + hex.EncodeToString(RandomBytes(24))
}

File: sdks/rust/src/crypto.rs
use hmac::{Hmac, Mac};
use sha2::{Digest, Sha256};

type HmacSha256 = Hmac<Sha256>;

/// Compute HMAC-SHA256 of a UTF-8 message using a UTF-8 secret and return hex.
pub fn hmac_sha256_hex(message: &str, secret: &str) -> String {
    let mut mac =
        HmacSha256::new_from_slice(secret.as_bytes()).expect("HMAC can take key of any size");
    mac.update(message.as_bytes());
    let result = mac.finalize();
    hex::encode(result.into_bytes())
}

/// Constant-time string comparison.
pub fn timing_safe_equal(a: &str, b: &str) -> bool {
    if a.len() != b.len() {
        return false;
    }
    let mut result: u8 = 0;
    for (x, y) in a.bytes().zip(b.bytes()) {
        result |= x ^ y;
    }
    result == 0
}

/// Generate cryptographically-secure random bytes.
pub fn random_bytes(length: usize) -> Vec<u8> {
    let mut buf = vec![0u8; length];
    getrandom::getrandom(&mut buf).expect("getrandom failed");
    buf
}

/// Encode bytes as lowercase hex string.
pub fn to_hex(bytes: &[u8]) -> String {
    hex::encode(bytes)
}

/// Decode hex string to bytes.
pub fn from_hex(hex_str: &str) -> Result<Vec<u8>, hex::FromHexError> {
    hex::decode(hex_str)
}

/// Compute SHA-256 of raw bytes and return as hex string.
pub fn sha256_hex(data: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(data);
    hex::encode(hasher.finalize())
}

/// Compute HMAC-SHA256 with raw byte key and message, returning raw bytes.
pub fn hmac_sha256_bytes(key: &[u8], message: &[u8]) -> Vec<u8> {
    let mut mac = HmacSha256::new_from_slice(key).expect("HMAC key");
    mac.update(message);
    mac.finalize().into_bytes().to_vec()
}

/// Generate a unique challenge ID (ch_ prefix + 16 random bytes as hex).
pub fn generate_id() -> String {
    format!("ch_{}", hex::encode(random_bytes(16)))
}

/// Generate a unique session token (st_ prefix + 24 random bytes as hex).
pub fn generate_session_token() -> String {
    format!("st_{}", hex::encode(random_bytes(24)))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_random_bytes_length() {
        assert_eq!(random_bytes(16).len(), 16);
        assert_eq!(random_bytes(32).len(), 32);
        assert_eq!(random_bytes(0).len(), 0);
    }

    #[test]
    fn test_to_hex_from_hex_roundtrip() {
        let original = random_bytes(32);
        let hex_str = to_hex(&original);
        let decoded = from_hex(&hex_str).unwrap();
        assert_eq!(original, decoded);
    }

    #[test]
    fn test_sha256_hex_known() {
        // SHA-256 of empty string
        let hash = sha256_hex(b"");
        assert_eq!(
            hash,
            "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
        );
        // SHA-256 of "hello"
        let hash2 = sha256_hex(b"hello");
        assert_eq!(
            hash2,
            "2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824"
        );
    }

    #[test]
    fn test_hmac_sha256_bytes_known() {
        let mac = hmac_sha256_bytes(b"key", b"message");
        assert_eq!(mac.len(), 32);
        let hex_str = to_hex(&mac);
        assert_eq!(
            hex_str,
            "6e9ef29b75fffc5b7abae527d58fdadb2fe42e7219011976917343065f58ed4a"
        );
    }

    #[test]
    fn test_hmac_sha256_hex_known() {
        let hex_str = hmac_sha256_hex("message", "key");
        // HMAC-SHA256("message", "key") where both are UTF-8 strings as keys
        assert!(!hex_str.is_empty());
        assert_eq!(hex_str.len(), 64);
    }

    #[test]
    fn test_generate_id_format() {
        let id = generate_id();
        assert!(id.starts_with("ch_"));
        assert_eq!(id.len(), 3 + 32); // "ch_" + 32 hex chars (16 bytes)
    }

    #[test]
    fn test_generate_session_token_format() {
        let token = generate_session_token();
        assert!(token.starts_with("st_"));
        assert_eq!(token.len(), 3 + 48); // "st_" + 48 hex chars (24 bytes)
    }

    #[test]
    fn test_timing_safe_equal_positive() {
        assert!(timing_safe_equal("hello", "hello"));
        assert!(timing_safe_equal("abc123", "abc123"));
    }

    #[test]
    fn test_timing_safe_equal_negative() {
        assert!(!timing_safe_equal("hello", "world"));
        assert!(!timing_safe_equal("abc", "abcd"));
        assert!(!timing_safe_equal("", "a"));
    }
}
